contracts/libraries/BytesMemory.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+/**
+ * @title A library for operating on bytes memory slices without copying.
+ */
+library BytesMemory {
+    error OutOfBounds();
+
+    /**
+     * @dev A struct representing a slice of bytes.
+     * This points directly to memory without copying the slice.
+     */
+    struct Slice {
+        uint256 pointer; // The pointer to the location of the slice in memory.
+        uint256 length; // The length of the slice in bytes.
+    }
+
+    /**
+     * @dev Creates a `Slice` from a bytes array.
+     * @param data The bytes array to create a slice from.
+     * @return A `Slice` struct representing the entire bytes array.
+     */
+    function wrap(bytes memory data) internal pure returns (Slice memory) {
+        uint256 pointer;
+        assembly ("memory-safe") { // solhint-disable-line no-inline-assembly
+            pointer := add(data, 0x20)
+        }
+
+        return Slice({
+            pointer: pointer,
+            length: data.length
+        });
+    }
+
+    /**
+     * @dev Returns a new `Slice` representing a portion of the original.
+     * @param data The original `Slice` to take a portion from.
+     * @param offset The offset in bytes from the start of the original `Slice`.
+     * @param size The size of the new `Slice` in bytes.
+     * @return A new `Slice` struct representing the specified portion of the original.
+     */
+    function slice(Slice memory data, uint256 offset, uint256 size) internal pure returns (Slice memory) {
+        if (offset + size > data.length) revert OutOfBounds();
+
+        return Slice({
+            pointer: data.pointer + offset,
+            length: size
+        });
+    }
+
+    /**
+     * @dev Converts a `Slice` back into a bytes array. The bytes array is returned without copying the data.
+     * @param piece The `Slice` to convert back to a bytes array.
+     * @return ret The bytes array represented by the `Slice`.
+     */
+    function unwrap(Slice memory piece) internal view returns (bytes memory ret) {
+        uint256 pointer = piece.pointer;
+        uint256 length = piece.length;
+        assembly ("memory-safe") { // solhint-disable-line no-inline-assembly
+            ret := mload(0x40)
+            mstore(0x40, add(0x20, length))
+            mstore(ret, length)
+
+            pop(staticcall(gas(), 0x04, pointer, length, add(ret, 0x20), length))
```

Comment:
We can shave some gas when copying short slices by direct mstore/mload instead of calling memcpy

Reply:
The following code used mload/mstore use `447` gas instead of `235` gas with 0x04 precompile contract
```
ret = new bytes(length);
assembly ("memory-safe") { // solhint-disable-line no-inline-assembly
     let dest := add(ret, 0x20)
     for { let i := 0 } lt(i, length) { i := add(i, 0x20) } {
         mstore(add(dest, i), mload(add(pointer, i)))
     }
 }
```

Reply:
Ok, if we don't use solidity to init `ret` it's better solution - `205` gas

Reply:
Are you referring to copying short segments, and does it matter to use different methods depending on the case? If so, what length should be used as a dividing criterion in such cases? Or should we add different implementations?

Reply:
On long slices we still need to use precompile

Reply:
At what byte count does a long slice begin?

Reply:
Ok, empirically it's 32 bytes, at 33 bytes it's already `279` > `235` with memcpy.